### ISSUE TITLE
FR: better error for ssh hostname changes

### ISSUE THREAD

**Issue by fowles** [martinvonz/jj#1507]
I recently tried to use `jj` to push to github.  The error I got was very perplexing to me.  In particular I saw this.

```
[sonmi]~/dev/protobuf$ jj git push
Branch changes to push to origin:
  Add branch test_jj to 5a9dbd9a4edc
Error: invalid or unknown remote ssh hostkey; class=Ssh (23); code=Certificate (-17)
Hint: Jujutsu uses libssh2, which doesn't respect ~/.ssh/config. Does `ssh -F /dev/null` to the host work?
```

I eventually traced this down to the recent github host changing

```
[sonmi]~/dev$ git clone git@github.com:fowles/abseil-hello.git
Cloning into 'abseil-hello'...
The authenticity of host 'github.com (140.82.114.3)' can't be established.
ED25519 key fingerprint is XXXXXXXXXXXXX
This key is not known by any other names.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'github.com' (ED25519) to the list of known hosts.
Warning: the ED25519 host key for 'github.com' differs from the key for the IP address '140.82.114.3'
Offending key for IP in /usr/local/google/home/kfm/.ssh/known_hosts:6
Are you sure you want to continue connecting (yes/no)? yes
```

so this fixed it, but an error message that told me something more (or better yet gave me the prompt that git does) would be awesome.

**Comment by necauqua** [comment]
While this is true and we need a better human error, the `ssh -F /dev/null` hint was added because of this - did it not help you trace it down to the github thing right away?

`ssh git@github.com` would fail with a giant MITM error, prompting you to do `ssh-key -R github.com` and then `ssh git@github.com` again to get the new fingerprint.

**Comment by fowles** [comment]
truthfully I did not try `ssh git@github.com` because I just assumed there is no possible way that github left the default port for ssh open in that way.  I am used to thinking of `ssh` solely in terms of machines I have remote terminal access to and I don't have remote terminal access to github.  I suspect that if the suggestion had said to try `ssh -F /dev/null git@github.com` I would have tried that.

The end result is that I actually spent about 45 minutes trying various stupid things before figuring out the workaround I did.

**Comment by fowles** [comment]
> Well what port would the git clone ~over ssh~ use then? :)

Honestly, I had never thought about it.  It is funny the abstractions we just accept :)

**Comment by clintonc** [comment]
Relatedly, when I attempted `jj git clone git@github.com:...`, I got the following error both before and after removing github.com from my known hosts:    

```
Error: Fetch failed: invalid or unknown remote ssh hostkey; class=Ssh (23); code=Certificate (-17)
```

When I did a normal `git clone` with the same argument, I got the normal new host flow:

```
The authenticity of host 'github.com (192.30.255.113)' can't be established.
ECDSA key fingerprint is SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'github.com,192.30.255.113' (ECDSA) to the list of known hosts.
```

In this case, there was no error message to help me figure out (or remediate) the problem.

**Comment by martinvonz** [comment]
If you install jj from head, you should at least get a hint there: https://github.com/martinvonz/jj/blob/0c36b1ad4e13f270b015868d1dd66cd42f11490e/src/commands/git.rs#L165-L181

I'm guessing `jj git clone` works now after the successful `git clone`?

**Comment by clintonc** [comment]
I installed jj from head (using the `cargo` command above), and a hint does show then. Cloning does not work afterwards (just hangs; verbose logs just shows a lot of lines like `INFO fetch{remote_name="origin" branch_name_globs=None git_settings=GitSettings { auto_local_branch: true }}: jujutsu_lib::git: using ssh_key_from_agent username="git"`, eventually failing with `Error: remote rejected authentication: Failed getting response; class=Ssh (23); code=Auth (-16)`) but that's a separate issue.

**Comment by zzhaolei** [comment]
(I turned on VPN)

I encountered a similar situation
https is ok, but ssh failed:
```shell
11:37 ❯ jj git clone https://github.com/zzhaolei/jj-demo.git
Fetching into new repo in "/Users/zhaolei/Study/jj-demo"
branch: main@origin [new] untracked
Setting the revset alias "trunk()" to "main@origin"
Working copy now at: srltlnwl d6627064 (empty) (no description set)
Parent commit      : luoyttxn 58cd234f main | add gitignore
Added 2 files, modified 0 files, removed 0 files
11:37 ❯ rm -rf jj-demo/
11:37 ❯ jj git clone git@github.com:zzhaolei/jj-demo.git
Fetching into new repo in "/Users/zhaolei/Study/jj-demo"
Error: failed to start SSH session: Failed getting banner; class=Ssh (23)
Hint: Jujutsu uses libssh2, which doesn't respect ~/.ssh/config. Does `ssh -F /dev/null` to the host work?
```

```shell
11:39 ❯ ssh -F /dev/null git@github.com
Connection closed by 198.18.1.134 port 22
```

git erverything is ok:
```shell
11:40 ❯ git clone https://github.com/zzhaolei/jj-demo.git
...
11:40 ❯ rm -rf jj-demo/
11:40 ❯ git clone git@github.com:zzhaolei/jj-demo.git
...
```

```
11:42 ❯ ssh -T git@github.com
Hi zzhaolei! You've successfully authenticated, but GitHub does not provide shell access.
```

`~/.ssh/config` and `~.gitconfig` has no configuration for `github.com`

**Comment by yuchenshi** [comment]
> I'm guessing `jj git clone` works now after the successful `git clone`?

@martinvonz I just ran into this when cloning for the first time on a new device. The `ssh -F /dev/null` hint was helpful and prompted me to accept the public key for github.com for the first time. And then it worked perfectly fine. Does `libssh2` have a way to add the key to known hosts (after user confirmation)?

**Comment by martinvonz** [comment]
> Does `libssh2` have a way to add the key to known hosts (after user confirmation)?

Yes, looks like it:  https://libssh2.org/libssh2_knownhost_addc.html There's also #3191 for switching from libssh2 to OpenSSH, which should solve lots of problems related to ssh config.

**Comment by steveklabnik** [comment]
Would there be an easy way to reproduce this issue? Now that we push directly via git, I'm guessing this is better, but I wanted to double check, but I'm not sure of the simplest way to change a hostname, heh.

**Comment by necauqua** [comment]
To get the confirmation prompt, you can delete the server line from `~/.ssh/known_hosts`, to get the giant scary warning that you're about to be hacked you can change the pubkeys in that file to something else (copy from some other server easiest) - or if you control the server you can ~~delete~~ back up the `/etc/ssh/ssh_host_*` files.

**Comment by steveklabnik** [comment]
Ah, right! So, funny fact: after I checked https://github.com/jj-vcs/jj/issues/6745#issuecomment-3607598918 by that method, I tried to find this issue again, but couldn't remember which one it was.

So, same thing: I believe that this one has been fixed, when I got rid of known_hosts, it prompted me, and I was able to select the options and they were respected.

**Comment by bnjmnt4n** [comment]
Closing since I believe this has been solved.
